### PR TITLE
Add "copy to clipboard" for all code blocks and "expand" buttons

### LINKED ISSUE #86851: Add copy button for code blocks
As discussed in https://github.com/rust-lang/rust/pull/86363, we could add a button allowing to copy the content of the source code. For this, I'll unify the content for the "run" button as well so it's not duplicated in case the run button is there as well. Or we could simply generate the button content when clicking on it using JS (because the "copy to clipboard" can only be done with JS), meaning we also generate the hidden lines (but keep them invisible).

Second point: for now the "run" button is always visible but its opacity goes to 1 when you hover it. I'll change this behaviour by hiding those buttons and only making them appear when hovering the code example so it improves the reading experience. However, I'll keep the opacity change behaviour when going on top of the button.

Last thing: should we add this copy button to all code blocks or only the rust ones?

cc @rust-lang/rustdoc

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #86851.

A little demo of the expand/collapse part:

https://user-images.githubusercontent.com/3050060/124787285-aa231f80-df48-11eb-898e-af85111da81e.mp4

Now some technical details: to prevent the copy of an `img` tag, I moved the image in the background of the button instead, allowing me to move it inside the CSS, preventing the size of the HTML to grow too much. By doing so however, it forced me to modify `rustdoc.css` in rustdoc directly when we generate the file. Nothing super complicated but still important enough to be noted.

The generated hidden lines are generated inside a `<span class="hidden">`, the JS handles the rest, it keeps the syntax highlighting of course. It forced me to update the `write_code` but nothing fancy either.

Another important detail: the "expand/collapse" button isn't generated if there is no hidden content.

All the new buttons are hidden in case the JS is disabled (`noscript.js`).

For the "copy code" feature, unfortunately, it cannot be tested (or at least I didn't find out how). So the only thing I can recommend is to try out locally. It's supposed to work as follow: only the visible code is copied. As simple as that. :)

Last information: I originally intended to hide the code blocks buttons (even the "run" one) by default and only make them appear when hovering the code block but I'll do that in another PR because I think this one is already big enough...

EDIT: Last information bis: In the JS code, the functions `copyCode` and `copy_path` are very similar, I intend to merge them even further in a later PR (I already started by creating the `copyContentToClipboard` function).

cc @jsha @Manishearth 
r? @jyn514

### CODE REVIEW

**Comment by jyn514** [comment]
I don't have time to review this right now. Is jsha still on vacation? They'd be a better reviewer.

**Comment by GuillaumeGomez** [comment]
They are in vacation indeed. I thought you were back, my bad. Feel free to assign someone else then. :)

**Comment by camelid** [comment]
The "expand" button looks to me like it would put the code in full-screen mode. mdBook already has a feature like this, so if we are going to implement it as well, it would probably be best to be consistent with mdBook's icon. Theirs is an eye icon ([link](https://doc.rust-lang.org/nightly/book/ch01-02-hello-world.html#anatomy-of-a-rust-program)):

image

**Comment by GuillaumeGomez** [comment]
@camelid Oh good idea! I'm gonna "steal" it from them. :3

**Comment by Manishearth** [comment]
I'd prefer the expanded code to be at least a little bit greyed out or something

**Comment by Manishearth** [comment]
TBH I'd rather wait for @jsha  to be back to give feedback on the UI.

File: src/librustdoc/html/highlight.rs [hunk 1]
```diff
@@ -57,20 +81,51 @@ fn write_header(out: &mut Buffer, class: Option<&str>, extra_content: Option<Buf
...
+    let mut iter = src.peekable();
+    let mut expand = false;
+
+    // For each `Line`, we replace DOS backlines with '\n'. This replace allows to fix how the code
+    // source with DOS backline characters is displayed.
+    while let Some(line) = iter.next() {
+        let (before, text, after) = match line {
+            Line::Hidden(text) => {
+                expand = true;
+                ("<span class=\"hidden\">", text.replace("\r\n", "\n"), "</span>")
```
Comment (Manishearth) [inline src/librustdoc/html/highlight.rs:94]: probably use a more specific class like hidden-code

Comment (GuillaumeGomez) [reply src/librustdoc/html/highlight.rs:94]: Good idea! That'll make it simpler for the background too.

**Comment by GuillaumeGomez** [comment]
> I'd prefer the expanded code to be at least a little bit greyed out or something

Good idea indeed!

> TBH I'd rather wait for @jsha to be back to give feedback on the UI.

Me too. I'll just keep it up-to-date until @jsha is back.

**Comment by Stupremee** [comment]
Just my opinion, but it looks kinda ugly to me that the Play button is so big and different from the other twos.

**Comment by GuillaumeGomez** [comment]
It's part of the things we need @jsha for. ;)

**Comment by notriddle** [comment]
While we're at it, could this be changed to use the play ▶ button for running the sample, like how mdbook does it?

**Comment by GuillaumeGomez** [comment]
Well, why not. :laughing: 

I'll also hide the buttons by default then.

**Comment by notriddle** [comment]
I'd still like to see https://github.com/rust-lang/rust/pull/86892#discussion_r664929566 done, and jsha to give feedback on the UI.

**Comment by GuillaumeGomez** [comment]
> I'd still like to see [#86892 (comment)](https://github.com/rust-lang/rust/pull/86892#discussion_r664929566) done, and jsha to give feedback on the UI.

After trying it out, there is no point in doing so since we already have a class for that after all... So I won't do it. And yes, better to wait for @jsha.

**Comment by jsha** [comment]
@GuillaumeGomez can you add instructions in the PR description on how to test locally? I've checked out the branch locally and run `cargo doc` against a project that has some hidden lines, but I don't see the added "copy to clipboard" and "expand" buttons.

**Comment by GuillaumeGomez** [comment]
@jsha I did nothing in particular, just built the std docs by doing `x.py doc std --stage 1` then:

Screenshot from 2021-09-28 16-40-03

Just search for `write_all_vectored` in the std docs and here you go!

**Comment by jsha** [comment]
Thanks! So, one piece of quick UI feedback: Right now we have one button that uses text ("Run") and two that are icons. We should use either all text or all icons in this location. My general preference is text over icons, but given there are three buttons, Run / Reveal / Copy would probably be too big, so we should make Run into a "play" icon like in mdBook.

Though, one other thought just to put it out there: We could change this to a small menu, like "tools", which would have Run / Reveal / Copy as menu items below it.

By the way, in https://github.com/rust-lang/rust/pull/86363 there was some confusion about why we don't have "Run" on _all_ documentation. I've felt that confusion myself a number of times. I think that's because the Run button disappears when it's not available, rather than being greyed out. If we grey out the Run button on non-std crates, that allows us to add a tooltip that says "Rust playground not available for this crate" or similarly explains the problem.

The Reveal button will only be present on a small subset of examples. Given that, it feels awkward to stick it in the middle of the two other icons. I have a slight preference for putting it all the way to the right, so the relative position of Run and Copy is always the same.

**Comment by GuillaumeGomez** [comment]
So, we have the following propositions:

 * If we keep the current display, change the "run" button text into an icon and grey it out when the example cannot be run and put the button on the right.
 * Move all the buttons into a menu (expanding on hover/click I assume). Should we use icons or text in this case? I assume text but not sure...

Anyway, which one seems preferable to you @jsha and to the others @rust-lang/rustdoc ?

**Comment by Nemo157** [comment]
I think 3 icons are small enough that a menu is unnecessary, and I'd prefer avoiding any hover states (partly selfishly, because I use vi-mode browser nav).

Making the run button appear greyed out could actually be great for getting crates to enable support for it, if the text was something like "Rust playground not _enabled_ for this crate" the crates that are lucky enough to be included in the playground might start adding  `#[doc(html_playground_url)]` attributes (alternatively, docs.rs might be able to auto-detect this and enable it for crates that are available 🤔).

**Comment by jsha** [comment]
> If we keep the current display, change the "run" button text into an icon and grey it out when the example cannot be run and put the button on the right.

Right now I think this is the preferable option. Note: I'm saying the "eye" (reveal) button should be on the right, not the "run" button.

> I'd prefer avoiding any hover states

I'm also generally anti-hover for menus. In https://github.com/rust-lang/docs.rs/pull/1081 I removed some hover menus from docs.rs, and linked to https://uxmovement.com/navigation/why-hover-menus-do-users-more-harm-than-good/ for details about why.

**Comment by Manishearth** [comment]
I also prefer avoiding hover menus

Though a thought: The eye button currently displays the state that one will reach by clicking it, not the current state. I know that UI design tends to be splti on this: i kinda prefer when it shows the _current_ state, so it shows an eye when stuff is visible, and a crossed eye when stuff is hidden.

**Comment by notriddle** [comment]
> I know that UI design tends to be splti on this: i kinda prefer when it shows the current state, so it shows an eye when stuff is visible, and a crossed eye when stuff is hidden.

I'm not picky about which option it goes with, *as long as its the same one as The Book* (which is rendered using mdBook). Which means it should be showing a crossed eye when stuff is visible, and a plain eye when stuff is hidden.

**Comment by jsha** [comment]
I agree on "same as The Book." Another useful question: are there other places where we should "current state" vs "state to be transitioned to" on a clickable thing? One example would be the `[-]` / `[+]` toggles, where we show "state to be transitioned to." Unfortunately that's that opposite of what The Book does. On balance I think "match The Book" wins.

**Comment by camelid** [comment]
> So, we have the following propositions:
> 
>   * If we keep the current display, change the "run" button text into an icon and grey it out when the example cannot be run and put the button on the right.
> 
>   * Move all the buttons into a menu (expanding on hover/click I assume). Should we use icons or text in this case? I assume text but not sure...

Personally, I would prefer we don't add these new buttons at all. Rustdoc's UI already has a lot of stuff going on, and having buttons on every code example adds even more visual components to the actual text of the docs. As I mentioned in https://github.com/rust-lang/rust/issues/86851#issuecomment-874349854, I don't feel like I understand what the use case of this feature is.

Part of why I don't understand the use case is that I find that most code examples I look at can't just be copied and pasted into my code; only a small piece, say a function call, is relevant, and that I can type myself. For example, with [`Iterator::filter()`](https://doc.rust-lang.org/nightly/std/iter/trait.Iterator.html#method.filter), most of the examples are just setup and assertions to demonstrate the function in context. Nothing except `.iter().filter()` would be relevant to my code, so I wouldn't copy-and-paste.

As discussed, mdBook also has these features, but, at least in mdBooks like The Book, the code samples are part of larger programs that the reader is building up, so it's more useful to have copy-to-clipboard in that situation. Whereas with rustdoc, code examples are usually not "real" programs but just demonstrations of the API in context.

I also want to note that a small technical downside to this feature is that it will increase page sizes, because now we need to include the hidden parts of examples in the page too.

---

But, if the consensus is to add this feature, I think I would lean towards having a three-vertical-dots–style menu, which, when clicked/tapped (not hovered), would show a list:

* Run
* Show hidden code
* Copy

I think having a menu opened by tapping an icon instead of three individual, potentially named, buttons would be less intrusive on the regular docs reading experience.

**Comment by Manishearth** [comment]
FWIW I am very ambivalent about adding this feature; I kinda agree with @camelid's position. But I'm not going to block it if most of the rest of the team wants it

**Comment by GuillaumeGomez** [comment]
Originally it was mostly about showing the hidden code in case the run button isn't available because people complained that it was fairly complicated for beginners to write the wrapping code (when `?` is used for example). I don't remember how we reached the point of adding a copy button though... I think that the "show hidden code" part is still a good idea (especially for beginners), even if it increases the code size (would be interesting to check by how much it increases it0.

**Comment by jyn514** [comment]
I think we should find a way to address that specific problem rather than a button to copy the whole code *including hidden lines*, that seems even more confusing.

**Comment by Manishearth** [comment]
In general I think we should be filing issues for _problems_, not _solutions_, and working from there.

**Comment by GuillaumeGomez** [comment]
Which is why I was suggesting to remove the add of the copy button and only keep the show/hide one.

**Comment by jyn514** [comment]
I'm going to close this for now untill we decide on the proper improvement; j don't think it will be exactly this PR.

**Pull request closed without merging**